--- usr.sbin/sysinstall/dist.c.orig	2010-10-26 05:38:14.000000000 -0700
+++ usr.sbin/sysinstall/dist.c	2010-10-29 07:24:46.000000000 -0700
@@ -79,6 +79,7 @@ static Distribution SrcDistTable[];
 static Distribution DistTable[] = {
     DTE_TARBALL("base",	    &Dists, BASE,     "/"),
     DTE_SUBDIST("kernels",  &Dists, KERNEL,   KernelDistTable),
+    DTE_TARBALL("perl",     &Dists, PERL,     "/"),
     DTE_TARBALL("doc",	    &Dists, DOCUSERLAND,      "/"),
     DTE_SUBDIST("docproj",  &Dists, DOC,      DocDistTable),
     DTE_TARBALL("games",    &Dists, GAMES,    "/"),
@@ -87,9 +88,7 @@ static Distribution DistTable[] = {
     DTE_TARBALL("proflibs", &Dists, PROFLIBS, "/"),
     DTE_TARBALL("dict",	    &Dists, DICT,     "/"),
     DTE_TARBALL("info",	    &Dists, INFO,     "/"),
-#ifdef __amd64__
     DTE_TARBALL("lib32",    &Dists, LIB32,    "/"),
-#endif
     DTE_SUBDIST("src",	    &Dists, SRC,      SrcDistTable),
     DTE_TARBALL("ports",    &Dists, PORTS,    "/usr"),
     DTE_TARBALL("local",    &Dists, LOCAL,    "/"),
