--- lib/libarchive/config_freebsd.h.orig	2012-03-02 22:15:13.000000000 -0800
+++ lib/libarchive/config_freebsd.h	2012-05-18 16:15:09.000000000 -0700
@@ -176,9 +176,4 @@
 #define	ARCHIVE_HASH_SHA256_OPENSSL 1
 #define	ARCHIVE_HASH_SHA384_OPENSSL 1
 #define	ARCHIVE_HASH_SHA512_OPENSSL 1
-#else
-#define	ARCHIVE_HASH_MD5_LIBC 1
-#define	ARCHIVE_HASH_SHA1_LIBC 1
-#define	ARCHIVE_HASH_SHA256_LIBC 1
-#define	ARCHIVE_HASH_SHA512_LIBC 1
 #endif
